pkg/common/nodeserver.go

```diff
@@ -3,16 +3,92 @@ package common
 import (
 	"context"
 	"errors"
+	"fmt"
 	"os"
 	"path/filepath"
 	"strings"
+	"time"
 
 	"github.com/container-storage-interface/spec/lib/go/csi"
+	"github.com/kubernetes-sigs/alibaba-cloud-csi-driver/pkg/metric"
 	"github.com/kubernetes-sigs/alibaba-cloud-csi-driver/pkg/utils"
+	"github.com/prometheus/client_golang/prometheus"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
 )
 
+func WrapNodeServerWithMetricRecorder(server csi.NodeServer, driverType string, client kubernetes.Interface) csi.NodeServer {
+	return &NodeServerWithMetricRecorder{
+		NodeServer: server,
+		driverType: driverType,
+		client:     client,
+	}
+}
+
+type NodeServerWithMetricRecorder struct {
+	csi.NodeServer
+	driverType string
+	client     kubernetes.Interface
+}
+
+func (s *NodeServerWithMetricRecorder) NodePublishVolume(ctx context.Context, req *csi.NodePublishVolumeRequest) (*csi.NodePublishVolumeResponse, error) {
+	ctx = withPodInfo(ctx, s.client, req)
+	resp, err := s.NodeServer.NodePublishVolume(ctx, req)
+	s.recordVolumeAttachmentTime(ctx, req, err)
+	return resp, err
+}
+
+func withPodInfo(ctx context.Context, client kubernetes.Interface, req *csi.NodePublishVolumeRequest) context.Context {
+	name, namespace := req.VolumeContext[podNameKey], req.VolumeContext[podNamespaceKey]
```

Comment:
extract duplicated codes into a new utils function

Reply:
I would move `withPodInfo` to util.go. Getter and setter should be in the same place. Maybe also make `PodInfo` private, to avoid any misuse. How about also cache the error into ctx? so that we will not trigger the error twice.

And maybe `withPodInfo` can return the updated ctx and the `*v1.Pod` at the same time, so that we can directly pass `*v1.Pod` into `recordVolumeAttachmentTime`